pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,304 @@
+
+	DefaultQueryCacheConfig = queryCacheConfig{
+		EvictionPolicy: FirstInFirstOut,
+		// TODO_MAINNET(@bryanchriswhite): Consider how we can "guarantee" good
+		// alignment between the TTL and the block production rate.
+		TTL: time.Minute,
+	}
+)
+
+// InMemoryCache provides a concurrency-safe in-memory cache implementation with
+// optional historical value support.
+type InMemoryCache[T any] struct {
+	config       queryCacheConfig
+	latestHeight atomic.Int64
+
+	itemsMu sync.RWMutex
+	// items type depends on historical mode:
+	// | historical mode |  type                          |
+	// | --------------- | ------------------------------ |
+	// | false           | map[string]cacheItem[T]        |
+	// | true            | map[string]cacheItemHistory[T] |
+	items map[string]any
+}
+
+// cacheItem wraps cached values with a timestamp for later comparison against
+// the configured TTL.
+type cacheItem[T any] struct {
+	value     T
+	timestamp time.Time
+}
+
+// cacheItemHistory stores cachedItems by height and maintains a sorted list of
+// heights for which cached items exist. This list is sorted in descending order
+// to improve performance characteristics by positively correlating index with age.
+type cacheItemHistory[T any] struct {
+	// sortedDescHeights is a list of the heights for which values are cached.
+	// It is sorted in descending order.
+	sortedDescHeights []int64
+	itemsByHeight     map[int64]cacheItem[T]
+}
+
+// NewInMemoryCache creates a new InMemoryCache with the configuration generated
+// by the given option functions.
+func NewInMemoryCache[T any](opts ...QueryCacheOptionFn) *InMemoryCache[T] {
+	config := DefaultQueryCacheConfig
+
+	for _, opt := range opts {
+		opt(&config)
+	}
+
+	return &InMemoryCache[T]{
+		items:  make(map[string]interface{}),
+		config: config,
+	}
+}
+
+// Get retrieves the value from the cache with the given key. If the cache is
+// configured for historical mode, it will return the value at the latest **known**
+// height, which is only updated on calls to SetAtHeight, and therefore is not
+// guaranteed to be the current height.
+func (c *InMemoryCache[T]) Get(key string) (T, error) {
+	if c.config.historical {
+		return c.GetAtHeight(key, c.latestHeight.Load())
+	}
+
+	c.itemsMu.RLock()
+	defer c.itemsMu.RUnlock()
+
+	var zero T
+
+	item, exists := c.items[key]
+	if !exists {
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
+	}
+
+	cItem := item.(cacheItem[T])
+	if c.config.TTL > 0 && time.Since(cItem.timestamp) > c.config.TTL {
+		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
+		// otherwise, the read lock would be insufficient. The value will be
+		// overwritten by the next call to Set().
+		return zero, ErrCacheMiss.Wrapf("key: %s", key)
+	}
+
+	return cItem.value, nil
+}
+
+// GetAtHeight retrieves the value from the cache with the given key, at the given
+// height. If a value is not found for that height, the value at the nearest previous
```

Comment:
> If a value is not found for that height, the value at the nearest previous height is returned.

Thoughts on adding a `nearestOk` bool to make this functionality ultra explicit?

Reply:
When I see a function like `GetAtHeight`, my immediate thought is that it'll return the value at the height provided (if available) or nil otherwise.

For `height=5`, it's the difference between `select * from table where height=5` and `select * from table where height <= 5`. 

There could be business logic that depends on state transitions AT A SPECIFIC HEIGHT, so I'm just afraid that it's not explicit enough.

Reply:
I take your point, and agree that his is a valid concern. However, I think we're getting hung up on an implementation detail that is actually an optimization. If the implementation were completely honest, it would be populating the history at **each** height, where the value would be the same as the previous height so long as it wasn't changed (or deleted; we have few of these objects: claims/proofs).

The fact that the heights are omitted here is an optimization which takes advantage of the assumption that this pattern holds due to the blockchain context that we're in. Even in the case of the few onchain objects which do violate this assumption , I don't think it becomes problematic. I also don't think we have a good reason to consider caching claims or proofs yet.

A concrete example would be something like:
1. A claim was created at height 10
2. The claim was queried and cached at height 12
3. The claim was settled and deleted at height 20
4. The claim was queried at height 30, for height 22
5. The claim would be returned even though it was deleted two blocks prior to the queried height

Reply:
```go
// HistoricalQueryCache extends QueryCache to support getting and setting values
// at multiple heights for a given key.
type HistoricalQueryCache[T any] interface {
	QueryCache[T]
	// GetAsOfHeight retrieves the nearest value <= the specified height
	GetAsOfHeight(key string, height int64) (T, error)
	// SetAsOfHeight adds or updates a value at a specific height
	SetAsOfHeight(key string, value T, height int64) error
}
```

Reply:
_tl;dr I still disagree but will leave the thread here :)_

There are "implicit assumptions"  in this conversation/thread.

Whenever that happens, I always think it's something that requires context or institutional knowledge. Further, it's something I can see us forgetting as well.

In either case, you are the main implementor and designer here. If this becomes an issue, I'm sure it'll resurface.